File: algorithms/algorithms-cplusplus/leetcode/Question_0842_Split_Array_into_Fibonacci_Sequence.cc
#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<int> splitIntoFibonacci(string num) {
    vector<int> res, cur;
    help(num, 0, cur, res);
    return res;
  }

 private:
  void help(string& num, int i, vector<int>& cur, vector<int>& res) {
    if (!res.empty()) {
      return;
    }
    if (i >= num.size() && cur.size() >= 3) {
      res = cur;
      return;
    }
    for (int j = i; j < num.size(); j++) {
      string s = num.substr(i, j - i + 1);
      if ((s.size() > 1 && s[0] == '0') || s.size() > 10) {
        break;
      }
      long n = stol(s), len = cur.size();
      if (n > INT_MAX) {
        break;
      }
      if (cur.size() >= 2 && n != (long)cur[len - 1] + cur[len - 2]) {
        continue;
      }
      cur.push_back(n);
      help(num, j + 1, cur, res);
      cur.pop_back();
    }
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0842_Split_Array_into_Fibonacci_Sequence.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.List;

public class Question_0842_Split_Array_into_Fibonacci_Sequence {
    public List<Integer> splitIntoFibonacci(String num) {
        List<Integer> res = new ArrayList<>(), cur = new ArrayList<>();
        help(num, 0, cur, res);
        return res;
    }

    private void help(String num, int i, List<Integer> cur, List<Integer> res) {
        if (!res.isEmpty()) {
            return;
        }
        if (i >= num.length() && cur.size() >= 3) {
            res.addAll(cur);
            return;
        }
        for (int j = i; j < num.length(); j++) {
            String s = num.substring(i, j + 1);
            if ((s.length() > 1 && s.charAt(0) == '0') || s.length() > 10) {
                break;
            }
            long n = Long.parseLong(s);
            int len = cur.size();
            if (n > Integer.MAX_VALUE) {
                break;
            }
            if (cur.size() >= 2 && n != (long) cur.get(len - 1) + cur.get(len - 2)) {
                continue;
            }
            cur.add((int) n);
            help(num, j + 1, cur, res);
            cur.remove(cur.size() - 1);
        }
    }
}
